Added dummy routetable for network policy only mode

juanfresia opened the pull request:
## Description

This PR adds a config flag for `felix` named `RouteSyncDisabled`:
* The default value is `false`.
* Setting this to `true` will prevent `felix` from doing any updates on the route table.
* This is applied to all instances of `RouteTable` (IPv4, IPv6, VXLAN and Wireguard).

In order to make sure no operations are done against the linux routing table, a new implementation of `RouteTable` called `DummyTable` was added. While using the `DummyTable` all operations are skipped, enabling the use of `felix` as a network policy controller only.

## Related issues/PRs

More information on the context for this change can be found in #5247.

<!-- If appropriate, include a link to the issue this fixes.
fixes #5247 
-->

## Release Note

```release-note
Added dummy routetable for network policy only mode
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calico?pullRequest=5454) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calico?pullRequest=5454) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calico?pullRequest=5454) it.</sub>

song-jiang commented:
Thanks @juanfresia for the PR. There are multiple places need to be updated in order to add a Felix configuration. 
- ./api directory. https://github.com/projectcalico/calico/blob/master/api/pkg/apis/projectcalico/v3/felixconfig.go and auto generated files
- ./libcalico directory.  CI and auto generated files

fasaxc commented on felix/dataplane/linux/int_dataplane.go:
```diff
@@ -709,9 +710,17 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
-		config.DeviceRouteSourceAddress, config.DeviceRouteProtocol, config.RemoveExternalRoutes, 0,
-		dp.loopSummarizer)
+	var routeTableV4 routeTable
+	routeTableV4 = &routetable.DummyTable{}
```
This only covers one instance of the RouteTable, there can be multiple:
* One for IPv4, one for IPv6
* One for VXLAN
* etc.

song-jiang commented on felix/config/config_params.go:
```diff
@@ -330,7 +330,8 @@ type Config struct {
 
-	RouteTableRange idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	RouteTableRange  idalloc.IndexRange `config:"route-table-range;1-250;die-on-fail"`
+	RouteSyncEnabled bool               `config:"bool;true"`
```
I would suggest to use `RouteSyncDisabled` with default value `false`. The configuration item is part of FelixConfiguration resource and should be omit-empty.

juanfresia commented:
Thanks you both for the comments, I'll send a commit to address them shortly

juanfresia commented:
Hi @song-jiang @fasaxc sorry for the delay. I've just updated the PR with the requested changes.
* I added the same change for `IPv6, VXLAN and Wireguard, which are all the uses of RouteTable I could find.
  * I think IPv6 and IPv4 are quite similar; but I'm not super familiar with the other two. If you know of any side effects of this change for those, you can point me in the right direction to fix them; or maybe it'd be an alternative to keep this flag from affecting VXLAN and Wireguard.
* I also changed the semantics to be `RouteSyncDisabled` with a `omitempty` config and a `false` default.
* `api` and `libcalico-go` sub-projects have been updated (hopefully, in the right way).

Also updated the PR's original description.

Let me know if there are any more needed changes/or ideas to be added here. Since this is basically adding a Dummy implementation of an interface, I'm not quite sure which tests would be valuable here; let me know if you have any ideas here as well.

song-jiang commented on felix/dataplane/linux/int_dataplane.go:
```diff
@@ -434,9 +435,16 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
-			dp.loopSummarizer)
+		var routeTableVXLAN routeTable
+		if !config.RouteSyncDisabled {
+			log.Info("RouteSyncDisabled is false.")
```
In most cases `RouteSyncDisabled` will be false so I suggest to use `log.Debug` . Similar suggestions apply to other RouteTable below.

juanfresia replied:
Sounds good, I'll change it for a `log.Debug` when we hit the default value for `RouteSyncDisabled`

song-jiang commented on felix/dataplane/linux/vxlan_mgr.go:
```diff
@@ -105,31 +105,47 @@ func newVXLANManager(
+		brt = &routetable.DummyTable{}
+	}
+
+	noEncapRTConstruct := func(interfaceRegexes []string, ipVersion uint8, vxlan bool, netlinkTimeout time.Duration,
```
Maybe move setting of `noEncapRTConstruct` to `if !dpConfig.RouteSyncDisabled` above?

song-jiang commented on felix/wireguard/route_table.go:
```diff
@@ -0,0 +1,24 @@
+	SetL2Routes(ifaceName string, targets []routetable.L2Target)
+	RouteRemove(ifaceName string, cidr ip.CIDR)
+	RouteUpdate(ifaceName string, target routetable.Target)
+}
```
I'm not sure why you need to add this file.

juanfresia replied:
So, Wireguard was using the `RouteTable` struct directly from the `routetable` package, and that meant that I could not override it directly with the `DummyTable`. All other uses of `RouteTable` went through the `routeTable` interface, from `intdataplane` package.

That interface did not include the `RouteRemove` and `RouteUpdate` methods `Wireguard` uses; and I didn't want to introduce a dependency between `wireguard` and `intdataplane` packages for this; so I replicated the interface with just the methods `wireguard` needs, checking that both `routetable.RouteTable` and `DummyTable` implements them.

That was my logic, but of course I could be missing something, do you have any suggestions on how would an alternative look like?

song-jiang commented on felix/wireguard/wireguard.go:
```diff
@@ -150,7 +150,7 @@ type Wireguard struct {
 
 	// Wireguard routing table and rule managers
-	routetable *routetable.RouteTable
+	routetable routeTable
```
Not sure about the change here.

juanfresia replied:
See the other comment, changed `routetable.RouteTable` struct with an `wireguard.routeTable` interface.

song-jiang commented:
@juanfresia I've reviewed your changes and added couple of comments. Thanks!

M00nF1sh commented on api/pkg/apis/projectcalico/v3/felixconfig.go:
```diff
@@ -370,6 +370,10 @@ type FelixConfigurationSpec struct {
 
+	// RouteSyncDisabled will disable all operations performed on the route table. Set to true to
+	// run in network-policy mode only.
+	RouteSyncDisabled *bool `json:"routeSuncDisabled,omitempty"`
```
typo: routeSuncDisabled-> routeSyncDisabled

M00nF1sh commented on libcalico-go/config/crd/crd.projectcalico.org_felixconfigurations.yaml:
```diff
@@ -496,6 +496,10 @@ spec:
                   information. - WorkloadIPs: use workload endpoints to construct
                   routes. - CalicoIPAM: the default - use IPAM data to construct routes.'
                 type: string
+              routeSuncDisabled:
```
typo: routeSuncDisabled-> routeSyncDisabled

juanfresia replied:
Thanks! Fixed now.

juanfresia commented:
@song-jiang Just fixed the log level as per your comment. I had left you some question regarding your other comments, what do you think would be the best alternative?

song-jiang commented:
@juanfresia I'm sorry I somehow missed this thread and have been on vacation lately. Thanks for all your work. 
I'd like to get input from @mikestephen on changes related to wireguard, but the rest code changes LGTM. 

By the way, could you please rebase on master branch again?

juanfresia commented:
Thanks @song-jiang , I'll rebase the changes

caseydavenport commented:
@juanfresia looks like this still needs a rebase.

caseydavenport commented:
/sem-approve

juanfresia commented:
Hi @mikestephen! I believe @song-jiang wanted your input on this change, do you have any thoughs on this?

song-jiang commented:
@juanfresia I had a chat with @mikestephen and @caseydavenport , we think at this point, the PR is good to get merged. I'll raise a follow up PR to address two improvements:
1. Fix Calico doc for the new

song-jiang commented:
@juanfresia I had a chat with @mikestephen and @caseydavenport , we think at this point, the PR is good to get merged. I'll raise a follow up PR to address two improvements:
1. Fix Calico doc for the new

caseydavenport commented:
/sem-approve

song-jiang commented:
/sem-approve

mgleung commented:
@juanfresia looks like this needs another rebase and then it's ready to go with the passing tests.

mgleung commented:
/sem-approve

juanfresia commented:
Thanks @mgleung, I'll sync it again against `master` shortly

caseydavenport commented:
/sem-approve

juanfresia commented:
Hi @song-jiang, you mentioned some documentation PRs missing; is there an example I can use as an example? Do you think we can merge this PR or should we keep it open until documentation is ready?

Other than that, is there anything we are missing to merge this change?

song-jiang commented:
@juanfresia Sorry I missed your comment. I think we can just merge this PR. We need to do the following:
1. Fix Calico doc for the new felix parameter.
2. Make wireguard start using the routeTable interface. 

I can take care both point but please let me know if you are happy to do the first one. Thanks.
